### PR TITLE
Fixed edge cases with Query Pairs

### PR DESCRIPTION
**Pull request by TWith2Sugars**

### CODE REVIEW

File: tests/Arachne.Uri.Tests/Types.Tests.fs [hunk 1]
```diff
@@ -147,4 +147,25 @@ let ``Uri Formatting/Parsing`` () =
...
+[<Test>]
+let ``Query Pairs with missing parameter value``() =
+    let expectedResult = Some ["param", Some("exists");"param1", None]
+    let query = Query.Query("param=exists&param1=")
+    let queryPairs = query |> fst Query.Pairs_
+    Assert.AreEqual(expectedResult, queryPairs)
+
+[<Test>]
+let ``Query Pairs with missing parameter value and equals sign``() =
+    let expectedResult = Some ["param", Some("exists");"param1", None]
```
Comment (panesofglass) [inline tests/Arachne.Uri.Tests/Types.Tests.fs]: I think the expected result here should be `Some ["param", Some("exists");"param1", Some("")]`. I've had cases in the past where I wanted an empty string in a query string. `None` should be reserved for the complete lack of a parameter in the query string.

File: tests/Arachne.Uri.Tests/Types.Tests.fs [hunk 2]
```diff
@@ -147,4 +147,25 @@ let ``Uri Formatting/Parsing`` () =
...
+    let query = Query.Query("param=exists&param1=alsoexists")
+    let queryPairs = query |> fst Query.Pairs_
+    Assert.AreEqual(expectedResult, queryPairs)
+
+[<Test>]
+let ``Query Pairs with missing parameter value``() =
+    let expectedResult = Some ["param", Some("exists");"param1", None]
+    let query = Query.Query("param=exists&param1=")
+    let queryPairs = query |> fst Query.Pairs_
+    Assert.AreEqual(expectedResult, queryPairs)
```
Comment (panesofglass) [inline tests/Arachne.Uri.Tests/Types.Tests.fs]: I think the expected result here should be `Some ["param", Some("exists");"param1", Some("")]`. I've had cases in the past where I wanted an empty string in a query string. `None` should be reserved for the complete lack of a parameter in the query string.

File: tests/Arachne.Uri.Tests/Types.Tests.fs [hunk 3]
```diff
@@ -147,4 +147,25 @@ let ``Uri Formatting/Parsing`` () =
...
+
+[<Test>]
+let ``Query Pairs with missing parameter value and equals sign``() =
+    let expectedResult = Some ["param", Some("exists");"param1", None]
+    let query = Query.Query("param=exists&param1")
+    let queryPairs = query |> fst Query.Pairs_
+    Assert.AreEqual(expectedResult, queryPairs)
```
Comment (panesofglass) [inline tests/Arachne.Uri.Tests/Types.Tests.fs]: This seems similar to the scenario where a parameter was missing entirely. Is this even a valid query string? I think `None` is the best option here, but I find it a strange case.

**Comment by panesofglass** [comment]
Do we have any tests for no query string at all? If not, would you mind adding one? I would expect that to return `None` and thread the `None` through the `Query.Pairs_` lens.

File: tests/Arachne.Uri.Tests/Types.Tests.fs (hunk 2 above)
Comment (TWith2Sugars) [inline tests/Arachne.Uri.Tests/Types.Tests.fs]: But the parameter does exists just the value doesn't. That was my line of thought with the "param1", None return type anyway.

File: tests/Arachne.Uri.Tests/Types.Tests.fs (hunk 3 above)
Comment (TWith2Sugars) [inline tests/Arachne.Uri.Tests/Types.Tests.fs]: From reading this: http://tools.ietf.org/html/rfc3986#section-3.4 it doesn't specify that you need '=', just that it's a common format

The appearance of a query param would indicate it's valid.

e.g. http://example.com/people?name=fred&isAlive

to me isAlive is a valid query parameter as in the example above it's been used as a flag.

**Comment by TWith2Sugars** [comment]
Added the extra test and changed to code to make it work :)

File: tests/Arachne.Uri.Tests/Types.Tests.fs (hunk 2 above)
Comment (panesofglass) [inline tests/Arachne.Uri.Tests/Types.Tests.fs]: I did not choose my words well. I was trying to make a note that returning `None` in both this and the case below does not provide a clear indication of how the query string was formatted. For example, I can't tell if `param1` was used as a flag or with a parameter value of `""`. I have seen APIs (consider HTML flag attributes) where multiple options were acceptable, e.g. `param1` (flag), `param1=callback` (provide a named value with the flag), etc. I think it best to leave these as separate cases and use `Some ""` for the case where the `=` was provided with no value.

File: tests/Arachne.Uri.Tests/Types.Tests.fs (hunk 3 above)
Comment (panesofglass) [inline tests/Arachne.Uri.Tests/Types.Tests.fs]: Thank you for referencing the spec. This works for me. I still find `None` an odd return value here. Perhaps `option<'T>` is not the right return value. Should we consider a custom DU for pairs?

``` fsharp
type QueryPair =
    | NameValue of string * string
    | Flag of string
    | Empty of string
```

Comment (TWith2Sugars) [inline tests/Arachne.Uri.Tests/Types.Tests.fs]: I'm happy either way. I'm not 100% sure how to do this in fparsec mind.

How about we push the current option type to at least fix this small bug then we can do the breaking change with the query pair DU.

File: tests/Arachne.Uri.Tests/Types.Tests.fs (hunk 2 above)
Comment (TWith2Sugars) [inline tests/Arachne.Uri.Tests/Types.Tests.fs]: Some "" for when = is provided is good with me. If you want I can try to make that change and do another push.

Comment (panesofglass) [inline tests/Arachne.Uri.Tests/Types.Tests.fs]: Yes, please. Hopefully, @kolektiv an chime in, too.

File: tests/Arachne.Uri.Tests/Types.Tests.fs (hunk 3 above)
Comment (panesofglass) [inline tests/Arachne.Uri.Tests/Types.Tests.fs]: Well, we could still avoid the DU if we just use `Some ""` for a `param1=`, I think.

**Comment by panesofglass** [comment]
Thank you again for pursuing this! You are a huge help!

File: tests/Arachne.Uri.Tests/Types.Tests.fs (hunk 3 above)
Comment (TWith2Sugars) [inline tests/Arachne.Uri.Tests/Types.Tests.fs]: I'll change the test/code to try and do that.

**Comment by TWith2Sugars** [comment]
It's a very useful library, glad to help!

**Comment by TWith2Sugars** [comment]
I'm not sure if pairPartV/pairPartP is the right name. I made the assumption that P = property and V = value though I'm probably wrong. Feel free to change this.
